File: Salao.java
import java.util.HashSet;
import java.util.Set;

public class Salao {

    // Directions used to check the 8 surrounding positions
    private static final int[][] DIRECTIONS = { 
        {-1, 0}, {1, 0}, {0, -1}, {0, 1}, 
        {-1, -1}, {-1, 1}, {1, -1}, {1, 1} 
    };

    // Verifies if a position is valid for placing a gang member
    public static boolean isPositionValid(String[][] board, int row, int column, String gangMember) {
        // Check surrounding cells for the same gang member
        for (int[] direction : DIRECTIONS) {
            int newRow = row + direction[0], newColumn = column + direction[1];
            if (newRow >= 0 && newRow < board.length && newColumn >= 0 && newColumn < board[0].length 
                && board[newRow][newColumn] == gangMember) {
                return false;
            }
        }

        // Check line of sight for gang members
        String enemyGang = (gangMember.equals("Bigode")) ? "Capeta" : "Bigode";

        for (int[] direction : DIRECTIONS) {
            int newRow = row + direction[0], newColumn = column + direction[1];
            while (newRow >= 0 && newRow < board.length && newColumn >= 0 && newColumn < board[0].length) {
                if (board[newRow][newColumn] == gangMember) {
                    return false; // Same gang member in line of sight
                } else if (board[newRow][newColumn] == enemyGang) {
                    break; // Enemy blocks the sight
                }
                newRow += direction[0];
                newColumn += direction[1];
            }
        }
        return true;
    }

    // Counts how many enemies are in the line of sight of the gunman
    public static int countEnemiesInSight(String[][] board, int row, int column, String gunman) {
        String enemyGang = (gunman == "Bigode") ? "Capeta" : "Bigode";
        int enemiesInSight = 0;

        for (int[] direction : DIRECTIONS) {
            int newRow = row + direction[0], newColumn = column + direction[1];
            while (newRow >= 0 && newRow < board.length && newColumn >= 0 && newColumn < board[0].length) {
                if (board[newRow][newColumn] == enemyGang) {
                    enemiesInSight++;
                    break;
                } else if (board[newRow][newColumn] != ".") {
                    break; // Blocked by another member
                }
                newRow += direction[0];
                newColumn += direction[1];
            }
        }
        return enemiesInSight;
    }

    // Check if every gunman on the board has at least two enemies in sight
    public static boolean isValidConfiguration(String[][] board) {
        for (int row = 0; row < board.length; row++) {
            for (int column = 0; column < board[0].length; column++) {
                if (board[row][column] == "Bigode" || board[row][column] == "Capeta") {
                    String gunman = board[row][column];
                    if (countEnemiesInSight(board, row, column, gunman) < 2) {
                        return false;
                    }
                }
            }
        }
        return true;
    }

    // Converts the current state of the board into a string
    public static String convertBoardToString(String[][] board) {
        StringBuilder sb = new StringBuilder();
        for (String[] row : board) {
            sb.append(String.join("", row)).append("\n");
        }
        return sb.toString();
    }

    // Backtracking to find all valid configurations
    public static void solve(String[][] board, int remainingB, int remainingC, Set<String> solutions, int startRow, int startcolumn) {
        // Base case: no more gang members to place
        if (remainingB == 0 && remainingC == 0) {
            if (isValidConfiguration(board)) {
                solutions.add(convertBoardToString(board)); // Store unique solution
            }
            return;
        }

        // Try placing gang members in available spots
        for (int row = startRow; row < board.length; row++) {
            for (int column = (row == startRow ? startcolumn : 0); column < board[0].length; column++) {
                if (board[row][column] == ".") {
                    // Place a member of gang B
                    if (remainingB > 0) {
                        board[row][column] = "Bigode";
                        if (isPositionValid(board, row, column, "Bigode")) {
                            solve(board, remainingB - 1, remainingC, solutions, row, column + 1);
                        }
                        board[row][column] = ".";  // Backtracking
                    }

                    // Place a member of gang C
                    if (remainingC > 0) {
                        board[row][column] = "Capeta";
                        if (isPositionValid(board, row, column, "Capeta")) {
                            solve(board, remainingB, remainingC - 1, solutions, row, column + 1);
                        }
                        board[row][column] = ".";  // Backtracking
                    }
                }
            }
        }
    }

    public static void main(String[] args) {
        long startTime = System.nanoTime();
        int n = Integer.parseInt(args[0]);
        int gangBCount = Integer.parseInt(args[1]);
        int gangCCount = Integer.parseInt(args[2]);
        
        String[][] board = new String[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                board[i][j] = ".";
            }
        }

        // Set to hold unique solutions
        Set<String> uniqueSolutions = new HashSet<>();

        // Start solving with backtracking
        solve(board, gangBCount, gangCCount, uniqueSolutions, 0, 0);
        long endTime = System.nanoTime();
        long executionTime = (endTime - startTime) / 1000000;

        // Output the results
        System.out.println("Number of possible solutions found: " + uniqueSolutions.size());
        System.out.println("Execution time: " + executionTime + "ms");
    }
}

File: salao.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_SIZE 20 // Adjust based on your needs
#define DIRECTIONS 8

typedef struct {
    char board[MAX_SIZE][MAX_SIZE];
    int occupied[MAX_SIZE][MAX_SIZE];
    int n;
} Salao;

int dir[DIRECTIONS][2] = {
    {-1, 0}, {1, 0}, {0, -1}, {0, 1},
    {-1, -1}, {-1, 1}, {1, -1}, {1, 1}
};

// Check if a position is within bounds
int isInBounds(int r, int c, int n) {
    return r >= 0 && r < n && c >= 0 && c < n;
}

// Check if placing a gunman of the given gang (B or C) at (row, col) is valid
int isValidPosition(Salao *s, int row, int col, char gang) {
    for (int i = 0; i < DIRECTIONS; i++) {
        int r = row + dir[i][0], c = col + dir[i][1];
        if (isInBounds(r, c, s->n) && s->occupied[r][c] && s->board[r][c] == gang) {
            return 0;
        }
    }

    char rival = (gang == 'B') ? 'C' : 'B';
    for (int i = 0; i < DIRECTIONS; i++) {
        int r = row + dir[i][0], c = col + dir[i][1];
        while (isInBounds(r, c, s->n)) {
            if (s->occupied[r][c] && s->board[r][c] == gang) {
                return 0; // Same gang member in line of sight
            } else if (s->occupied[r][c] && s->board[r][c] == rival) {
                break; // Rival blocks the sight
            }
            r += dir[i][0];
            c += dir[i][1];
        }
    }
    return 1;
}

// Count how many rival gang members the gunman at (row, col) is "aiming" at
int countRivalsAimedAt(Salao *s, int row, int col, char gang) {
    char rival = (gang == 'B') ? 'C' : 'B';
    int aimingCount = 0;

    for (int i = 0; i < DIRECTIONS; i++) {
        int r = row + dir[i][0], c = col + dir[i][1];
        while (isInBounds(r, c, s->n)) {
            if (s->occupied[r][c] && s->board[r][c] == rival) {
                aimingCount++;
                break;
            } else if (s->occupied[r][c]) {
                break; // Blocked by a teammate
            }
            r += dir[i][0];
            c += dir[i][1];
        }
    }
    return aimingCount;
}

// Check the entire board to ensure all gunmen are aiming at at least two rivals
int isValidStandoff(Salao *s) {
    for (int row = 0; row < s->n; row++) {
        for (int col = 0; col < s->n; col++) {
            if (s->occupied[row][col]) {
                char gunman = s->board[row][col];
                if (countRivalsAimedAt(s, row, col, gunman) < 2) {
                    return 0; // Invalid standoff
                }
            }
        }
    }
    return 1; // Valid standoff
}

// Convert the board to a string for easy comparison and storing in a set
void boardToString(Salao *s, char *result) {
    for (int row = 0; row < s->n; row++) {
        for (int col = 0; col < s->n; col++) {
            result[row * s->n + col] = s->board[row][col];
        }
    }
    result[s->n * s->n] = '\0'; // Null-terminate
}

// The main backtracking function
void solve(Salao *s, int bCount, int cCount, int *solutionCounter, char uniqueSolutions[][MAX_SIZE * MAX_SIZE], int *uniqueCount) {
    if (bCount == 0 && cCount == 0) {
        if (isValidStandoff(s)) {
            char boardStr[MAX_SIZE * MAX_SIZE + 1];
            boardToString(s, boardStr);

            // Check for uniqueness
            for (int i = 0; i < *uniqueCount; i++) {
                if (strcmp(uniqueSolutions[i], boardStr) == 0) {
                    return; // Already counted
                }
            }

            // Store the unique solution
            strcpy(uniqueSolutions[*uniqueCount], boardStr);
            (*uniqueCount)++;
            (*solutionCounter)++;
        }
        return;
    }

    for (int row = 0; row < s->n; row++) {
        for (int col = 0; col < s->n; col++) {
            if (!s->occupied[row][col]) {
                // Try placing a gang B member
                if (bCount > 0) {
                    s->board[row][col] = 'B';
                    s->occupied[row][col] = 1;
                    if (isValidPosition(s, row, col, 'B')) {
                        solve(s, bCount - 1, cCount, solutionCounter, uniqueSolutions, uniqueCount);
                    }
                    s->occupied[row][col] = 0; // Backtrack
                }

                // Try placing a gang C member
                if (cCount > 0) {
                    s->board[row][col] = 'C';
                    s->occupied[row][col] = 1;
                    if (isValidPosition(s, row, col, 'C')) {
                        solve(s, bCount, cCount - 1, solutionCounter, uniqueSolutions, uniqueCount);
                    }
                    s->occupied[row][col] = 0; // Backtrack
                }
                s->board[row][col] = '.'; // Clear the cell
            }
        }
    }
}

int main(int argc, char *argv[]) {
    if (argc != 4) {
        printf("Usage: %s <n> <bCount> <cCount>\n", argv[0]);
        return 1;
    }

    int n = atoi(argv[1]);
    int bCount = atoi(argv[2]);
    int cCount = atoi(argv[3]);

    Salao salao;
    salao.n = n;
    memset(salao.board, '.', sizeof(salao.board));
    memset(salao.occupied, 0, sizeof(salao.occupied));

    int solutionCounter = 0;
    char uniqueSolutions[100000][MAX_SIZE * MAX_SIZE + 1]; // Adjust size as needed
    int uniqueCount = 0;

    solve(&salao, bCount, cCount, &solutionCounter, uniqueSolutions, &uniqueCount);

    printf("Total number of unique solutions: %d\n", solutionCounter);
    return 0;
}